arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2893,6 +2920,319 @@ void RestReplicationHandler::handleCommandLoggerTickRanges() {
+          TRI_ASSERT(response.isOpenArray());
+          TRI_ASSERT(subOpen);
+          subOpen = false;
+          response.close();
+          TRI_ASSERT(response.isOpenArray());
+          resumeNext = range.at(1).getNumber<std::size_t>() + 1;
+          ++current;
+          if (current < body.length()) {
+            range = body.at(current);
+          }
+        }
+      }
+      if (subOpen) {
+        TRI_ASSERT(response.isOpenArray());
+        response.close();
+        subOpen = false;
+        TRI_ASSERT(response.isOpenArray());
+      }
+      TRI_ASSERT(!subOpen);
+      TRI_ASSERT(response.isOpenArray());
+      // exit scope closes "ranges"
+    }
+
+    TRI_ASSERT(response.isOpenObject());
+
+    if (body.length() > 0 &&
+        resumeNext <= body.at(body.length() - 1).at(1).getNumber<std::size_t>()) {
+      response.add("resume", VPackValue(resumeNext));
+    }
+  }
+
+  generateResult(rest::ResponseCode::OK, response.slice());
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief return the requested documents from a given collection, if
+///        available
+/// @response VPackArray, containing VPackObject documents or errors
+//////////////////////////////////////////////////////////////////////////////
+
+void RestReplicationHandler::handleCommandRevisionDocuments() {
+  RevisionOperationContext ctx;
+  if (!prepareRevisionOperation(ctx)) {
+    return;
+  }
+
+  bool success = false;
+  VPackSlice const body = this->parseVPackBody(success);
+  if (!success) {
+    // error already created
+    return;
+  }
+  bool badFormat = !body.isArray();
+  if (!badFormat) {
+    for (std::size_t i = 0; i < body.length(); ++i) {
+      if (!body.at(i).isNumber()) {
+        badFormat = true;
+        break;
+      }
+    }
+  }
+  if (badFormat) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  "body needs to be an array of revisions");
+    return;
+  }
+
+  std::size_t constexpr sizeLimit = 16 * 1024 * 1024;
+  std::size_t size = 0;  // running total, approximation
+  RevisionReplicationIterator& it =
+      *static_cast<RevisionReplicationIterator*>(ctx.iter.get());
+
+  VPackBuilder response;
+  {
+    VPackArrayBuilder docs(&response);
+
+    for (std::size_t i = 0; i < body.length(); ++i) {
+      TRI_voc_rid_t rev = body.at(i).getNumber<TRI_voc_rid_t>();
+      if (it.hasMore() && it.revision() < rev) {
+        it.seek(rev);
+      }
+      VPackSlice res =
+          it.hasMore() ? it.document() : velocypack::Slice::emptyObjectSlice();
+      if (size + res.byteSize() > sizeLimit && !response.slice().isEmptyArray()) {
+        break;
+      }
+      response.add(res);
+      size += res.byteSize();
+    }
+  }
+
+  HttpResponse* httpResponse = dynamic_cast<HttpResponse*>(_response.get());
+  if (httpResponse == nullptr) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "invalid response type");
+  }
```

Comment:
Should we make this work with VST as well? Right now it seems like it wouldn't work.
However, to the best of my knowledge, @graetzer has fixed all other places in the code that were hard-coded to a single protocol, so we shouldn't introduce new ones with this PR.

Reply:
jup you can just move in the VPackBuffer ?